### ISSUE TITLE
Broker crashes if there is an empty log segment

### ISSUE THREAD

**Issue by deepthidevaki** [camunda-cloud/zeebe#6979]
**Describe the bug**

[logs](https://console.cloud.google.com/logs/viewer?expandAll=false&timestamp=2021-05-04T14%3A09%3A56.909000000Z&project=camunda-cloud-240911&minLogLevel=0&customFacets&limitCustomFacetWidth=true&interval=JUMP_TO_TIME&scrollTimestamp=2021-05-04T09%3A42%3A12.919247000Z&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%22de990736-75b6-4e2b-8b51-65b2ccd9174d-zeebe%22%0Aresource.labels.cluster_name%3D%22ultrachaos%22%0Aresource.labels.pod_name%3D%22zeebe-0%22%0Aresource.labels.project_id%3D%22camunda-cloud-240911%22%0Aresource.labels.location%3D%22europe-west1-d%22%0Aresource.labels.container_name%3D%22zeebe%22%0Atimestamp%3D%222021-05-04T09%3A42%3A10.873656000Z%22%0AinsertId%3D%2295xz4qg4ckwrmf%22)

```
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[atomix-utils-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.nio.Buffer.createPositionException(Unknown Source) ~[?:?]
	at java.nio.Buffer.position(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.position(Unknown Source) ~[?:?]
	at java.nio.MappedByteBuffer.position(Unknown Source) ~[?:?]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.reset(MappedJournalSegmentWriter.java:202) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.<init>(MappedJournalSegmentWriter.java:66) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.createWriter(JournalSegment.java:151) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.<init>(JournalSegment.java:62) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.newSegment(SegmentedJournal.java:461) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegment(SegmentedJournal.java:472) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegments(SegmentedJournal.java:520) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:228) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:85) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:148) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.log.RaftLogBuilder.build(RaftLogBuilder.java:149) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:233) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:155) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:257) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:231) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:175) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:145) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:103) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:113) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:201) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
	at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:203) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.lambda$start$1(DefaultPartitionService.java:80) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
	at java.util.concurrent.ConcurrentHashMap$ValueSpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.start(DefaultPartitionService.java:81) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.core.Atomix.lambda$startServices$0(Atomix.java:178) ~[atomix-1.0.0-alpha7.jar:1.0.0-alpha7]
```
I inspected disk of the broker.  The segment that caused the problem had a file length of 0. So even the `JournalSegmentDescriptor` was not written to the file.
However what is more strange was that the file was "raft-partition-partition-5-0.log". The id of the segment is 0 which should never happen. The segment id's start from 1 always.

**Expected behavior**

* There should not be any segment with id 0.
* If it is the last segment, then it is most likely empty because the broker crashed before writing anything to the file. Hence it would be safe to ignore that segment and create a new one.

**Comment by MiguelPires** [comment]
**TLDR**: a partial write caused a zero-length file to be loaded into a buffer which was then used with the SBE decoder to read the fields. Since the buffer was empty, all of the fields were read as zero. This is why there was a file 'raft-partition-partition-5-0.log' even though no segment should have a 0 id and why we failed while reading from an empty buffer.

Long version:
Looking into the data folder a few weird things stand out. First, file 'raft-partition-partition-5-0.log' has an id of zero which should never happen since the initial id is 1 and they're always incremented. Second, the same file has a file size of zero:

```
root@zeebe-0:/usr/local/zeebe/data/raft-partition/partitions/5# ls -lh
total 257M
drwxr-xr-x 2 root root 4.0K May  4 09:42 pending
-rw-r--r-- 1 root root    0 May  4 09:42 raft-partition-partition-5-0.log
-rw-r--r-- 1 root root 128M May  4 09:33 raft-partition-partition-5-368.log
-rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-369.log
-rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-370.log
-rw-r--r-- 1 root root   78 May  3 13:00 raft-partition-partition-5.conf
-rw-r--r-- 1 root root  256 May  7 14:33 raft-partition-partition-5.meta
drwxr-xr-x 2 root root  16K May  4 09:35 runtime
drwxr-xr-x 3 root root 4.0K May  4 09:31 snapshots
```

At first, the empty file pointed to a partial write and the zero id pointed to corruption of a previous file. However, the key thing to note is that the last log segment 'raft-partition-partition-5-370.log' may seem to have the full 128MB but its actual allocated size is 0:

```
root@zeebe-0:/usr/local/zeebe/data/raft-partition/partitions/5# ls -lhs 
total 257M
4.0K drwxr-xr-x 2 root root 4.0K May  4 09:42 pending
   0 -rw-r--r-- 1 root root    0 May  4 09:42 raft-partition-partition-5-0.log
128M -rw-r--r-- 1 root root 128M May  4 09:33 raft-partition-partition-5-368.log
128M -rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-369.log
   0 -rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-370.log
4.0K -rw-r--r-- 1 root root   78 May  3 13:00 raft-partition-partition-5.conf
4.0K -rw-r--r-- 1 root root  256 May  7 14:49 raft-partition-partition-5.meta
 16K drwxr-xr-x 2 root root  16K May  4 09:35 runtime
4.0K drwxr-xr-x 3 root root 4.0K May  4 09:31 snapshots
```

So the partial write actually occurred in 'raft-partition-partition-5-370.log'. The timeline of events is the following:

On 04/05/2021 - 11:36, zeebe-0 crashed while creating a log segment with id 370. This is supported by the fact that the last log before the restart is at 11:35:58 CEST and the time of creation of the file is 09:36:51 UTC. It is also supported by the fact that its real size is 0:
```
root@zeebe-0:/usr/local/zeebe/data/raft-partition/partitions/5# ls -lhs --time-style=full-iso
total 257M
4.0K drwxr-xr-x 2 root root 4.0K 2021-05-04 09:42:08.182905902 +0000 pending
   0 -rw-r--r-- 1 root root    0 2021-05-04 09:42:10.871392216 +0000 raft-partition-partition-5-0.log
128M -rw-r--r-- 1 root root 128M 2021-05-04 09:33:55.268323034 +0000 raft-partition-partition-5-368.log
128M -rw-r--r-- 1 root root 128M 2021-05-04 09:36:51.043576425 +0000 raft-partition-partition-5-369.log
   0 -rw-r--r-- 1 root root 128M 2021-05-04 09:36:51.051576891 +0000 raft-partition-partition-5-370.log
4.0K -rw-r--r-- 1 root root   78 2021-05-03 13:00:56.074183086 +0000 raft-partition-partition-5.conf
4.0K -rw-r--r-- 1 root root  256 2021-05-07 14:49:21.696284442 +0000 raft-partition-partition-5.meta
 16K drwxr-xr-x 2 root root  16K 2021-05-04 09:35:53.141198840 +0000 runtime
4.0K drwxr-xr-x 3 root root 4.0K 2021-05-04 09:31:52.532163536 +0000 snapshots
```

On 04/05/2021 - 11:39, broker zeebe-0 restarts. When loading segment 'raft-partition-partition-5-370.log', we allocate a buffer and read the empty file into it.
https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java#L507
However, we don't check if any bytes were actually read so at this point we happily proceed with an empty buffer. Then we use it to create the segment descriptor object and, since we don't check that the SBE schema id, the template id and schema version match, we don't detect that this buffer is empty and the descriptor is successfully created.

https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptor.java#L68

Then we use the in-memory descriptor to load it again (I'm not sure why, might be a left over from before). However, since the descriptor was made by wrapping an empty buffer, when reading its fields we can only read 0. So when we create the segment file in `loadSegment`, we create it with the id 0. 

https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java#L463

When the creating the actual JournalSegment object, we map the segment file into memory. However, the length of the mapped region is given also by the descriptor so this is also set to 0.
https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegment.java#L59

Lastly, we create the segment writer which calls reset to advance its buffer's position past the descriptor bytes.
https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/MappedJournalSegmentWriter.java#L199

Since we used length 0 when mapping the segment file, the limit of the mapped buffer is zero and the call to `position` fails because we've passed a position larger than the limit:
```
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[atomix-utils-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.nio.Buffer.createPositionException(Unknown Source) ~[?:?]
	at java.nio.Buffer.position(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.position(Unknown Source) ~[?:?]
	at java.nio.MappedByteBuffer.position(Unknown Source) ~[?:?]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.reset(MappedJournalSegmentWriter.java:202) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.<init>(MappedJournalSegmentWriter.java:66) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.createWriter(JournalSegment.java:151) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.<init>(JournalSegment.java:62) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.newSegment(SegmentedJournal.java:461) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegment(SegmentedJournal.java:472) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegments(SegmentedJournal.java:520) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:228) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:85) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:148) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.log.RaftLogBuilder.build(RaftLogBuilder.java:149) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:233) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:155) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:257) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:231) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:175) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:145) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:103) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:113) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:201) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
	at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:203) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.lambda$start$1(DefaultPartitionService.java:80) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
        ....
```

There are several checks that we can add (I've tried to point them out in the previous paragraphs) that would've made this issue easier to diagnose. However, they wouldn't have resolve the issue because the root cause is a broker crash causing a partially written log segment. This raises the same issue that we've encountered before, namely, that we need to be able to distinguish partial writes that can be ignored/deleted from actual corruption that we must preserved and dealt with manually to avoid losing acknowledged data. 

One solution can be writing a zero byte as the first byte of the descriptor (currently we just write a version 1 as a placeholder). Then we can write the actual descriptor, flush it and then write the descriptor version as the starting magic byte. This way, when reading from the file, we can tell if was fully written by looking at the magic byte. The good news is that, since we already write this byte, we can do this in a backwards compatible way after the release (good thing too since it's 17:59 on the Friday before the release)

**Comment by Zelldon** [comment]
Very nice summarized @MiguelPires thanks! :rocket: 

> The good news is that, since we already write this byte, we can do this in a backwards compatible way after the release (good thing too since it's 17:59 on the Friday before the release)

Love this part, especially the last one :laughing:

**Comment by MiguelPires** [comment]
> Very nice summarized @MiguelPires thanks! 

Thanks Chris!

The solution described in my previous comment would work for now but it has two disadvantages: 1) it requires an extra write+flush and 2) it would have to change when deal with corruption at the descriptor. If the magic byte became corrupted then we couldn't distinguish between a partial and corruption at the descriptor. While this would be fine for now, since a partial write is much more likely than descriptor corruption, we'll eventually want a solution for both so we might as well have it in mind now. A more robust solution would be to just use the `lastWrittenIndex`, that we already write to the metastore, to tell if the segment contains entries that must be preserved or if it can be discarded. If we can't read a descriptor, we'd just look at the stored index and compare it to the index of the previous segment (we need to sort segment before instead of after). If the metastore's index is higher than the previous segment's index, then it contained data that can't be deleted immediately.

**Comment by deepthidevaki** [comment]
It would be good if we can reuse `lastWrittenIndex`  :+1:

> (we need to sort segment before instead of after).

Can we sort them before reading the segment descriptor?

**Comment by MiguelPires** [comment]
We can use the id part of the file name to sort them.
